### PR TITLE
Support Boot Order configuration

### PR DESCRIPTION
**Pull request by lsun100**
This commit adds support to boot order configuration in config
file /etc/bf.cfg. Each entry BOOT<N> has the following format:
  PXE:
    BOOT<N> = NET-<NIC_P0 | NIC_P1 | OOB | RSHIM>-<IPV4 | IPV6>
  UEFI Shell:
    BOOT<N> = UEFI_SHELL
  DISK: boot entries created during OS installation.
    BOOT<N> = DISK

The example below configures PXE boot over the 2nd ConnectX port.
It fails, it'll fall through the default booting sequence from
disk (like eMMC).
  BOOT0 = NET-NIC_P1-IPV4
  BOOT1 = DISK

### CODE REVIEW

File: bfcfg [hunk 1]
```diff
@@ -27,16 +27,26 @@
...
-bfcfg_version=0.1
+bfcfg_version=0.2
 cfg_file=/etc/bf.cfg
 log_file=/tmp/bfcfg.log
 dump_mode=0
 
+efivars=/sys/firmware/efi/efivars
+efi_global_var_guid=8be4df61-93ca-11d2-aa0d-00e098032b8c
+efi_vlan_var_guid=9e23d768-d2f3-4366-9fc3-3a7aba864374
+rshim_mac_sysfs=${efivars}/RshimMacAddr-${efi_global_var_guid}
```
Comment (kblaiech) [inline bfcfg:38]: Perhaps, we should use "rshim_efi_mac_sysfs" to be consistent with "oob_efi_mac_sysfs"?

File: bfcfg [hunk 2]
```diff
@@ -188,6 +202,353 @@ part_info()
...
+{
+  boot_cfg_add_header $1
+  boot_cfg_add_len $1 48
+  boot_cfg_add_name $1 "EFI Internal Shell"
+  printf \
+"\\x01\\x03\\x18\\x00\\x0b\\x00\\x00\\x00\\x00\\xf0\\x4f\\xfe\\x00\\x00\\x00\\x00"\
+"\\x7f\\x62\\xa6\\xfe\\x00\\x00\\x00\\x00\\x04\\x06\\x14\\x00\\x83\\xa5\\x04\\x7c"\
+"\\x3e\\x9e\\x1c\\x4f\\xad\\x65\\xe0\\x52\\x68\\xd0\\xb4\\xd1" >> $1
+  boot_cfg_add_tail $1
+}
```
Comment (kblaiech) [inline bfcfg:348]: Note sure if this is valid across bootloaders, the shell device path may change

File: bfcfg [hunk 3]
```diff
@@ -188,6 +202,353 @@ part_info()
...
+
+#
+# Boot Entry configuration
+# Each entry BOOT<N> could have the following format:
+#   PXE:
+#     BOOT<N> = NET-<NIC_P0 | NIC_P1 | OOB | RSHIM>-<IPV4 | IPV6>
```
Comment (kblaiech) [inline bfcfg:354]: vlan parameter is missing?

File: bfrec [hunk 4]
```diff
@@ -94,10 +94,10 @@ do
 done
 
 if [ -n "$bootctl_mode" ] && [ -n "$capsule_mode" ]; then
-    cat <<EOF
-ERROR: Use either options '--bootctl' or '--capsule'
-EOF
+    echo "ERROR: Use either options '--bootctl' or '--capsule'"
     exit 1
+elif [ -z "$bootctl_mode" ] && [ -z "$capsule_mode" ]; then
+    bootctl_mode=1
```
Comment (kblaiech) [inline bfrec:100]: add 'option_name="bootctl"'?

File: bfmisc [hunk 5]
```diff
@@ -36,16 +36,6 @@ set -e
 
 psid_boot_override()
```
Comment (kblaiech) [inline bfmisc:37]: Maybe we should remove this function?

File: bfcfg (hunk 1 above)
Comment (lsun100) [reply bfcfg:38]: done

File: bfcfg (hunk 2 above)
Comment (lsun100) [reply bfcfg:348]: Good catch! Another edk2 code review request has been sent to address this issue. It creates or fix the UEFI shell entries. (https://bu-gerrit2.mtbu.labs.mlnx/#/c/edk2/+/6127/)

File: bfcfg (hunk 3 above)
Comment (lsun100) [reply bfcfg:354]: Done

File: bfrec (hunk 4 above)
Comment (lsun100) [reply bfrec:100]: Done

File: bfmisc (hunk 5 above)
Comment (lsun100) [reply bfmisc:37]: Done

File: bfmisc [hunk 6]
```diff
@@ -33,22 +33,6 @@ set -e
 # This script executes miscellaneous tasks. These are vendor/customer
 # specific operations that are required during installation procedure.
 ##
```
Comment (kblaiech) [inline bfmisc:35]: Q: does it make sense to keep this script? I think the bcfg approach gives us more flexibility than this script, right?

Comment (lsun100) [reply bfmisc:35]: Done. Removed

**Pull request merged**